maple-core/src/reactive.rs

```diff
@@ -141,78 +188,78 @@ where
-
-    out
+    create_selector_with(derived, |_, _| false)
 }
 
 /// Creates a memoized value from some signals. Also know as "derived stores".
-pub fn create_memo<F, Out>(derived: F) -> StateHandle<Out>
+/// Unlike [`create_memo`], this function will not notify dependents of a change if the output is the same.
+/// That is why the output of the function must implement [`PartialEq`].
+///
+/// To specify a custom comparison function, use [`create_selector_with`].
+pub fn create_selector<F, Out>(derived: F) -> StateHandle<Out>
 where
     F: Fn() -> Out + 'static,
-    Out: Clone + 'static,
+    Out: PartialEq + 'static,
 {
-    let derived = Rc::new(derived);
-    let (memo, set_memo) = create_signal(None);
-
-    create_effect({
-        let derived = derived.clone();
-        move || {
-            set_memo(Some(derived()));
-        }
-    });
-
-    // return memoized result
-    let memo_result = move || Rc::new(Option::as_ref(&memo()).unwrap().clone());
-    Rc::new(memo_result)
+    create_selector_with(derived, PartialEq::eq)
 }
 
 /// Creates a memoized value from some signals. Also know as "derived stores".
 /// Unlike [`create_memo`], this function will not notify dependents of a change if the output is the same.
-/// That is why the output of the function must implement `PartialEq`.
-pub fn create_selector<F, Out>(derived: F) -> StateHandle<Out>
+///
+/// It takes a comparison function to compare the old and new value, which returns `true` if they
+/// are the same and `false` otherwise.
+///
+/// To use the type's [`PartialEq`] implementation instead of a custom function, use
+/// [`create_selector`].
+pub fn create_selector_with<F, Out, C>(derived: F, comparator: C) -> StateHandle<Out>
 where
     F: Fn() -> Out + 'static,
-    Out: Clone + PartialEq + std::fmt::Debug + 'static,
+    Out: 'static,
+    C: Fn(&Out, &Out) -> bool + 'static,
 {
-    let derived = Rc::new(derived);
-    let (memo, set_memo) = create_signal(None);
-
-    create_effect({
-        let derived = derived.clone();
-        let memo = memo.clone();
-        move || {
-            let new_value = Some(derived());
-            if *untracked(|| memo()) != new_value {
-                set_memo(new_value);
+    enum EffectState<T: 'static> {
+        Uninitialized(NonNull<Option<StateHandle<T>>>),
+        Initialized(Signal<T>),
+        Temporary,
+    }
+
+    let mut handle = None;
+    let state = Cell::new(EffectState::Uninitialized(NonNull::from(&mut handle)));
+
+    create_effect(move || {
+        let memo = match state.replace(EffectState::Temporary) {
+            // We haven't been called yet, so initialize the signal.
+            EffectState::Uninitialized(mut handle_ptr) => {
+                let memo = Signal::new(derived());
+
+                // SAFETY: This is the first time the function is called, and so this pointer is
+                // still valid.
+                *unsafe { handle_ptr.as_mut() } = Some((*memo).clone());

```

Comment:
Not sure why `unsafe` is needed here. Couldn't we just mutate `handle` directly?

Reply:
It is necessary there as the closure needs to be `'static` (so it can't borrow local variables), but I'm working on something that will remove the need for `unsafe` now.